src/lib/optimizer/join_ordering/greedy_operator_ordering.cpp

```diff
@@ -0,0 +1,167 @@
+
+    if (edge.vertex_set.none()) {
+      uncorrelated_predicates.insert(uncorrelated_predicates.end(), edge.predicates.begin(), edge.predicates.end());
+    } else {
+      remaining_edge_indices.emplace_back(edge_idx);
+    }
+  }
+
+  /**
+   * 1.3 Initialize "plan_by_edge". At all times, for each (uncorrelated) edge, we hold a plan at hand that would be
+   *       produced when joining the vertex clusters that this edge connects.
+   *       We need this plan both to
+   *         -> determine the edge with the lowest cardinality
+   *         -> set the plan associated with the newly formed vertex cluster, when joining the clusters with the lowest
+   *          cardinality
+   *
+   */
+  auto plan_by_edge = std::vector<PlanCardinalityPair>{join_graph.edges.size()};
+  for (const auto& edge_idx : remaining_edge_indices) {
+    plan_by_edge[edge_idx] = _build_plan_for_edge(join_graph.edges[edge_idx], vertex_clusters);
+  }
+
+  /**
+   * 2. Main loop of the algorithm.
+   *      During each iteration, we select the edge that has a minimum cardinality and join the vertex clusters that it
+   *      connects
+   */
+  while (!remaining_edge_indices.empty()) {
+    /**
+     * 2.1 Find the edge with the lowest cardinality and remove it from "remaining_edge_indices"
+     */
+    std::sort(remaining_edge_indices.begin(), remaining_edge_indices.end(),
+              [&](const auto& lhs, const auto& rhs) { return plan_by_edge[lhs].second > plan_by_edge[rhs].second; });
+
+    const auto lowest_cardinality_edge_idx = remaining_edge_indices.back();
+    remaining_edge_indices.pop_back();
+
+    const auto& lowest_cardinality_edge = join_graph.edges[lowest_cardinality_edge_idx];
+
+    /**
+     * 2.2 Determine the vertex set "joined_vertex_set" of the new cluster created by joining all clusters connected by
+     *      the edge with the lowest cardinality.
+     *      While doing so, remove all the individual vertex clusters connected by the edge from the set of
+     *      vertex_clusters - we will add a new vertex cluster comprised of all of them afterwards
+     */
+    auto joined_vertex_set = JoinGraphVertexSet{join_graph.vertices.size()};
+    for (auto vertex_cluster_iter = vertex_clusters.begin(); vertex_cluster_iter != vertex_clusters.end();) {
+      if ((vertex_cluster_iter->first & lowest_cardinality_edge.vertex_set).any()) {
+        joined_vertex_set |= vertex_cluster_iter->first;
+        vertex_cluster_iter = vertex_clusters.erase(vertex_cluster_iter);
+      } else {
+        ++vertex_cluster_iter;
+      }
+    }
+
+    /**
+     * 2.3 Add a new vertex cluster consisting of all the vertex clusters joined by the edge with the lowest cardinality
+     */
+    vertex_clusters.emplace(joined_vertex_set, plan_by_edge[lowest_cardinality_edge_idx].first);
+
+    /**
+     * 2.4 Update the plans associated with all edges that connect to the newly created cluster.
+     */
+    for (const auto& edge_idx2 : remaining_edge_indices) {
```

Comment:
You probably already tried to find a better name here?

Reply:
Nope, but have done so now.